12.0 imp commown devices extract quant and picking replacement in contract

9184Lucky opened the pull request.

Commit:
Add relation between lots and contracts.

The relations relies on the relation between pickings, scraps and
contracts.

Commit:
Remove quants and picking from contract

Commit:
Remove unused function

Commit:
Add link between contracts and move line

Also improve a test on website_sale_b2b to check contract to lot link

Commit:
Add explicit test for action_scrap

Commit:
Modify common picking function to return move line instead of picking

We also add the possibility to reuse a picking but this is not used for
now. It will be used later.

fcayre commented on commown_devices/models/crm_lead.py:
```diff
@@ -44,14 +44,14 @@ def _check_picking_on_stage_change(self):
         "Validate shipping"
         super(CrmLead, self).delivery_perform_actions()
-        picking = self.contract_id.picking_ids.filtered(_assigned)
+        picking = self.contract_id.move_ids.mapped("picking_id").filtered(_assigned)
```
ça donne envie de créer une méthode `pending_picking` sur le contrat : 
```
def pending_pickings(self):
    return self.move_ids.mapped("picking_id").filtered(_assigned)
```
(qui présuppose ici que la fonction _assigned est au bon endroit)

9184Lucky replied:
Fait, vu que la fonction assigned se retrouve utilisée dans crm_lead et dans contract je l'ai déplacée dans common

fcayre commented on commown_devices/models/stock_move.py:
```diff
@@ -0,0 +1,44 @@
+    contract_id = fields.Many2one("contract.contract", string="Contract")
+
+    def _action_confirm(self, merge=True, merge_into=False):
+        if self.env.context.get("dont_merge_moves", False):
```
ajouter une docstring

9184Lucky replied:
Fait

fcayre commented on commown_devices/models/stock_move.py:
```diff
@@ -0,0 +1,44 @@
+        return res
+
+    def update_lot_location(self):
+        for rec in self:
```
Ajouter une docstring

9184Lucky replied:
Fait

fcayre commented on commown_devices/models/stock_move.py:
```diff
@@ -0,0 +1,44 @@
+
+    def update_lot_location(self):
+        for rec in self:
+            lot_ids = rec.move_line_ids.mapped("lot_id")
```
Je pense (sans preuve) qu'il est plus efficace d'écrire : `rec.mapped("move_line_ids.lot_id")` qui effectue sans doute une seule requête SQL.

9184Lucky replied:
Fait

fcayre commented on commown_devices/models/stock_move.py:
```diff
@@ -0,0 +1,44 @@
+                lot_nb = len(lot_ids)
+                if lot_nb > 1:
+                    _logger.warning(
+                        "%s lots found on move %s. Associating %s new lot to contract %s",
```
new lot > new lots

9184Lucky replied:
Fait

fcayre commented on commown_devices/views/contract.xml:
```diff
@@ -32,16 +21,16 @@
               <tree
-                  colors="red: state not in ('done', 'cancel'); grey: state == 'cancel'"
-                  default_order="date_done ASC, id ASC"
+                  colors="red:move_id.scrapped"
```
ça j'ai du mal à croire que ça fonctionne : `move_id` ne doit pas être définie ici, `scrapped` tout simplement ça ne fonctionne pas ? De plus je crois que colors ne fonctionne plus en v12 et qu'il faut utiliser `decoration-muted` et `decoration-error` par exemple sauf erreur de ma part.

fcayre replied:
cette remarque reste valable, sauf si corrigé plus loin...

fcayre commented on commown_devices/views/contract.xml:
```diff
@@ -2,17 +2,6 @@
-      </xpath>
-    </field>
-  </record>
-
```
pourquoi tu supprimes cette section dans ce commit ? Pas clair pour moi car le champ quant_nb n'est pas supprimé dans ce commit j'ai l'impression. Pas grave non plus mais bizarre.

fcayre commented on commown_devices/models/contract.py:
```diff
@@ -45,6 +45,8 @@ class Contract(models.Model):
         string="Stock Move",
     )
 
+    lot_ids = fields.One2many("stock.production.lot", "contract_id", string="Lots")
```
Typo dans le message du 2ème commit : bewtween > between (que ce con de github ne permet pas de commenter au bon endroit...)

fcayre commented on commown_devices/models/common.py:
```diff
@@ -81,6 +81,75 @@ def find_products_orig_location(env, products, stocks=None, compute_summary=Fals
 
 
+def create_move_from_lots(picking, located_lots):
+    """Create a stock move each lot. Do not réuse existing move because we might want to
```
Typos dans la docstring:
- "move each lot" > "move for each lot" je suppose
- "réuse" > "reuse"

9184Lucky replied:
Fait

fcayre commented on commown_devices/models/common.py:
```diff
@@ -81,6 +81,75 @@ def find_products_orig_location(env, products, stocks=None, compute_summary=Fals
+    The stock.move.line are automatically created when the picking is assigned"""
+    env = picking.env
+    moves_by_lot = {}
+    for lot in located_lots.keys():
```
l'appel à `keys()` est inutile dans une boucle sur un dictionnaire : ça boucle bien sur les clefs si on ne met rien.
Ça pourrait valoir le coup d'indiquer dans la docstring le format de located_lots (je suppose {lot: location} ; ah non, d'après la suite :smile: ).

9184Lucky replied:
Alors fait pour le .keys(), et pour le dictionnaire en fait il s'aggit d'un dictionnaire {lot: {"loc": location}} la raison de ceci est que je voulais garder une structure similaire à located_products qui est un dictionnaire {product: {"loc": location, "qty": quantity}} mais ce choix est discutable, qu'en penses tu ?

fcayre commented on commown_devices/models/common.py:
```diff
@@ -81,6 +81,75 @@ def find_products_orig_location(env, products, stocks=None, compute_summary=Fals
+    return moves_by_products
+
+
+def update_picking_datas(picking, from_locations, dest_location, origin):
```
data est indénombrable, ne prend jamais de "s"

Une docstring me paraît nécessaire ici.

9184Lucky replied:
Fait

fcayre commented on commown_devices/models/common.py:
```diff
@@ -89,9 +158,11 @@ def internal_picking(
 ):
-    """Create picking with tracked and untracked products"""
+    """Create picking with tracked and untracked products, if a picking is passed as an
+    argument it wil try to update the existing picking with new move_lines"""
```
"argument it wil try to update the existing picking with new move_lines" > "argument, try to update the existing picking with new move lines"

9184Lucky replied:
Fait

fcayre commented on commown_devices/models/common.py:
```diff
@@ -81,6 +81,75 @@ def find_products_orig_location(env, products, stocks=None, compute_summary=Fals
+    picking.origin += ", " + origin
+
+
 def internal_picking(
```
Ne faut-il pas renommer cette fonction ?

9184Lucky replied:
Certainement, dans une version je l'avait renommée "create_or_update_internal_picking". Qu'en penses tu ?

fcayre commented on commown_devices/models/picking.py:
```diff
@@ -7,7 +7,16 @@
     _inherit = "stock.picking"
 
-    contract_id = fields.Many2one("contract.contract", string="Contract")
+    contract_ids = fields.Many2many(
```
c'est même un champ `related="move_lines.contract_id"` ça devrait éviter la méthode compute inutile et l'option nocreate dans la vue (car c'est alors carrément readonly par défaut je pense).

9184Lucky replied:
Fait

fcayre commented on commown_devices/models/contract.py:
```diff
@@ -129,31 +129,3 @@ def _create_or_reuse_picking(
-
-    @api.multi
-    def stock_at_date(self, date=None):
-        "Return the lots at partner's location at the given date"
```
good catch !

9184Lucky replied:
:)

Commit:
Move _assigned function to common as it will later be used elsewhere

Commit:
Fix error in test class inheritance

Commit:
Revert unused reuse_picking feature

Commit:
Migrate picking and scrap moves to contract

Commit:
Shorten location name on contract view

Commit:
Fix link between contract and scrap's move

When a scrap is created, it has no move. Therefore we need to store the contract on the scrap and associate it to its move when created

Commit:
Add a toggle to hide move lines without lot in contract view

Commit:
Add a test to prevent scrapping an already scrapped device

Commit:
Improve move line display on contract view

Add decorations depending on line in or out destination

Commit:
Add test to improve coverage

Commit:
Remove a unused line in tests

fcayre commented on commown_devices/models/contract.py:
```diff
@@ -43,6 +56,13 @@ def _compute_quant_ids(self):
+
+    def _compute_move_line_ids(self):
+        for rec in self:
+            rec.move_line_ids = rec.move_ids.mapped("move_line_ids")
```
related="move_ids.move_line_ids" dans le One2many fait la même chose en plus clair et court je pense .

9184Lucky replied:
:+1:

fcayre commented on commown_devices/models/project_task.py:
```diff
@@ -78,9 +78,9 @@ def _may_be_related_lots(self):
-            ).mapped("move_line_ids.lot_id")
+            lots |= contract.move_line_ids.filtered(
+                lambda ml: ml.picking_id.state == "assigned"
+            ).mapped("lot_id")
```
utiliser _assigned ?

9184Lucky replied:
Fait

fcayre commented on commown_devices/models/stock_move.py:
```diff
@@ -0,0 +1,51 @@
+    contract_id = fields.Many2one("contract.contract", string="Contract")
+
+    def _action_confirm(self, merge=True, merge_into=False):
+        """Override the methode so moves are merged or not depending on the context"""
```
Typo : methode > method

fcayre commented on commown_devices/models/stock_move.py:
```diff
@@ -0,0 +1,51 @@
+        return super()._action_confirm(merge=merge, merge_into=merge_into)
+
+    def _action_done(self):
+        """Overrride the methode to trigger update_lot_location methode"""
```
Même typo : methode > method

fcayre commented on commown_devices/models/stock_move.py:
```diff
@@ -0,0 +1,51 @@
+        between lot and contract depending on the move destination.
+        """
+        for rec in self:
+            lot_ids = rec.mapped("move_line_ids.lot_id")
```
perso j'aime pas nommer "*_id" ou "*_ids" une variable qui contient des entités, j'utilise ces noms pour les champs (ça a un peu plus de sens car en base de données ce sont des entiers, même si je trouve ça nul) ou pour des... ids (= des entiers).

9184Lucky replied:
Ok, j'ai changé, je comprends ce que tu veux dire. Moi je faisait ça pour essayer de coller à la syntaxe odoo qui appelle _ids des champs qui contienent des entités (même si en base elles sont stockés en entier tu peux appeler les méthodes du modèles et les mettre à jour).

fcayre commented on commown_devices/models/stock_move.py:
```diff
@@ -0,0 +1,51 @@
+                    lot_ids.update({"contract_id": rec.contract_id})
+                elif (
+                    rec.location_id == contract_loc
+                    and lot_ids.mapped("contract_id") == rec.contract_id
```
Je pense qu'on devrait ici lever une erreur direct si le contrat est le bon mais pas la localisation de départ, sinon j'ai peur qu'on s'enfonce dans des bugs de données de pire en pire.

9184Lucky replied:
A voir ensemble, pas sur de voir ce que tu veux dire !

fcayre replied:
Fait

fcayre commented on commown_devices/models/stock_picking.py:
```diff
@@ -12,7 +12,17 @@
+    @api.depends("move_lines.contract_id")
+    def _compute_contract_ids(self):
+        for record in self:
+            record.contract_ids = self.move_lines.mapped("contract_id")
```
Un related="move_lines.contract_id" est plus clair et ne requiert pas de code donc de test.

9184Lucky replied:
:+1:

fcayre commented on rental_fees/models/rental_fees_computation.py:
```diff
@@ -15,7 +15,7 @@
 def _move_contract(move_line):
     "Helper to return the contract related to a device move"
-    return move_line.move_id.picking_id.contract_id
+    return move_line.move_id.contract_id
```
Cette fonction n'est pas utilisée apparemment, tu peux la virer et ainsi améliorer la couverture apparente de ton patch !

9184Lucky replied:
:+1:

fcayre commented on commown_devices/models/stock_move_line.py:
```diff
@@ -0,0 +1,16 @@
+
+    is_contract_in = fields.Boolean(compute="_compute_is_contract_in", store=False)
+
+    def _compute_is_contract_in(self):
```
ça devrait être très rapide de rajouter un appel à is_contract_in dans un test pour améliorer la couverture, ce qui me semble important ici.

9184Lucky replied:
Il ya a déjà un test (dans test_stock_move_line.py), juste que codecov l'as pas pris en compte !!

Commit:
i18n fr10n

fcayre committed:
Robustify the code that sets the contract/ lot relation on move validation

We now use a location hierarchy (based on their partners' hierarchy) to
circumvent problems due to partners changing from B2C to B2B or the
like.

Also, in the migration script, log an error instead of raising when move
and contract locations are inconsistent

Commit:
Fix migration script

The post-migrate script was looping on the contract's moves no matter
the move's global chronological order.

9184Lucky commented:
/ocabot merge patch
